rich/syntax.py

```diff
@@ -690,6 +743,81 @@ def _get_syntax(
                     yield from wrapped_line
                     yield new_line
 
+    def _apply_stylized_ranges(self, *, code: str, text: Text) -> None:
+        """
+        Apply stylized ranges to a text instance,
+        using the given code to determine the right portion to apply the style to.
+
+        Args:
+            code (str): Code to highlight.
+            text (Text): Text instance to apply the style to.
+        """
+        # N.B. using "\n" here is much faster than using metacharacters such as "^" or "\Z":
+        newlines_offsets = [
+            match.start() for match in re.finditer("\n", code, flags=re.MULTILINE)
+        ]
+        # Let's add outer boundaries at each side of the list:
+        newlines_offsets.insert(0, -1)
+        newlines_offsets.append(len(code))
+
+        for stylized_range in self._stylized_ranges:
+            start, end = _get_code_indexes_for_syntax_positions(
+                newlines_offsets, (stylized_range.start, stylized_range.end)
+            )
+            if start is not None and end is not None:
+                text.stylize(stylized_range.style, start, end)
+
+    def _process_code(self, code: str) -> Tuple[bool, str]:
+        """
+        Applies various processing to a raw code string
+        (normalises it so it always ends with a line return, dedents it if necessary, etc.)
+
+        Args:
+            code (str): The raw code string to process
+
+        Returns:
+            Tuple[bool, str]: the boolean indicates whether the raw code ends with a line return,
+                while the string is the processed code.
+        """
+        ends_on_nl = code.endswith("\n")
+        processed_code = code if ends_on_nl else code + "\n"
+        processed_code = (
+            textwrap.dedent(processed_code) if self.dedent else processed_code
+        )
+        processed_code = processed_code.expandtabs(self.tab_size)
+        return ends_on_nl, processed_code
+
+
+def _get_code_indexes_for_syntax_positions(
+    newlines_offsets: Sequence[int], positions: Sequence[SyntaxPosition]
```

Comment:
Does `positions` need to be a sequence?

AFAICT we only ever call this for 2 positions. I think it may make sense for this function to take a single `SyntaxPosition`, and the caller can call it twice.